### PR TITLE
Set cgroup values via libcontainer

### LINKED ISSUE #88197: kubelet incorrectly configures `kubepods.slice` systemd unit when kube-reserved and/or system-reserved is used
**What happened**:

Kubelet is started on CentOS 7 with systemd with the following switches (among others) `--cgroup-driver=systemd`, `--enforce-node-allocatable=pods`, `--kube-reserved=cpu=100m,ephemeral-storage=2Gi,memory=1.5Gi`

With these settings kubelet correctly creates `kubepods.slice` cgroup with memory limit set to total RAM on the node (~4Gi in this example) minus kube-reserved (1.5Gi in this example)
```
# free -b
              total        used        free      shared  buff/cache   available
Mem:     3973447680   167432192  3489378304      528384   316637184  3584466944

# cat /sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes
2362834944
```

At the same time corresponding systemd unit `kubepods.slice` created by kubelet has a `MemoryLimit` property incorrectly set to total RAM of the node (without subtracting kube-reserved amount):
```
# systemctl show kubepods.slice
Slice=-.slice
ControlGroup=/kubepods.slice
...
MemoryLimit=3973447680
...

# systemctl cat kubepods.slice
...
# /run/systemd/system/kubepods.slice.d/50-MemoryLimit.conf
[Slice]
MemoryLimit=3973447680
...
```

This leads to systemd sometimes resetting the limit in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` cgroup file to total RAM on the node without accounting for the kube-reserved (and/or system-reserved) amount.
This in turn can lead to pods using more memory than expected and OOM kill events, and unstable node behavior in general

**What you expected to happen**:

`kubepods.slice` systemd unit `MemoryLimit` property should be set to the same value as `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` taking into account kube-reserved and system-reserved amounts.

When kubelet s restarted with different kube-reserved and/or system-reserved values, both `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit should be changed to the same value even if the cgroups and the systemd unit already exist.

**How to reproduce it (as minimally and precisely as possible)**:

Run kubelet on any systemd linux with the switches specified above and observe the discrepancy between memory limits defined in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
  1.16.0, but also applies to all earlier and later versions
- Cloud provider or hardware configuration:
   virtualbox VM, but applies to any infrastructure
- OS (e.g: `cat /etc/os-release`):
   centos 7, but applies to any systemd based linux flavor
- Kernel (e.g. `uname -a`):
   `3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux` but applies to any other kernel
- Install tools:
   kublr, but applies to any other tool
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by odinuge**
**What type of PR is this?**

/kind bug
/kind cleanup

**What this PR does / why we need it**:

WIP to run all e2e tests.

Currently we set cgroup values in `pkg/kubelet/cm/cgroup_manager_linux.go` manually per subsystem, howerever, when using the systemd driver, we should set them via opencontainers/runc/libcontainer to make sure we update the systemd slices and scopes.

Will come back with more info when I have tested more.

**Which issue(s) this PR fixes**:

Fixes #88197

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fix resource enforcement when using systemd cgroup driver
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by odinuge** [comment]
Still having some issues with cgroup v2 that need to be addressed. However, cc:

/cc @rphillips @olegch

**Comment by odinuge** [comment]
/retest

flakes

**Comment by harche** [comment]
Thanks @odinuge, I cherry picked this change on top of https://github.com/kubernetes/kubernetes/pull/101888 and in my local setup the stats summary API test seems to be passing fine. :)

**Comment by odinuge** [comment]
> @odinuge, I cherry picked this change on top of&nbsp;#101888&nbsp;and in my local setup the stats summary API test seems to be passing fine. :)

Yeah, as I said yesterday in https://github.com/kubernetes/kubernetes/issues/102092#issuecomment-843136063, this will fix the issues with runc-rc94

The change is however quite high risk, so I would like CI signal from `node-kubelet-serial` (and possiblly the same with systemd) in order to make sure this isn't introducing any regressions. I do not have time/resources to run that testing manually for the next few days tho., since I am busy with my uni work.. :/

**Comment by harche** [comment]
@odinuge no problem, thanks again for your PR. In my testing I ran entire `NodeConformance` and `NodeFeature` tests  (not just stats summary test) and it passed fine.

**Comment by odinuge** [comment]
> @odinuge no problem, thanks again for your PR. In my testing I ran entire NodeConformance and NodeFeature tests (not just stats summary test) and it passed fine.

Awesome! Thanks!

/cc @liggitt @kolyshkin @dims @bobbypage

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 1]
```diff
@@ -577,10 +517,11 @@ func (m *cgroupManagerImpl) Update(cgroupConfig *CgroupConfig) error {
 	}
 
 	unified := libcontainercgroups.IsCgroup2UnifiedMode()
+	var paths map[string]string
 	if unified {
 		libcontainerCgroupConfig.Path = m.Name(cgroupConfig.Name)
 	} else {
-		libcontainerCgroupConfig.Paths = m.buildCgroupPaths(cgroupConfig.Name)
```
Comment (odinuge) [inline pkg/kubelet/cm/cgroup_manager_linux.go:583]: Because of:
https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/fs/fs.go#L284-L286

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 2]
```diff
@@ -351,32 +351,6 @@ func getSupportedSubsystems() map[subsystem]bool {
 	return supportedSubsystems
 }
 
-// setSupportedSubsystemsV1 sets cgroup resource limits on cgroup v1 only on the supported
-// subsystems. ie. cpu and memory. We don't use libcontainer's cgroup/fs/Set()
-// method as it doesn't allow us to skip updates on the devices cgroup
```
Comment (odinuge) [inline pkg/kubelet/cm/cgroup_manager_linux.go:356]: This is now resolved by setting `Resources.SkipDevices = true`.

**Comment by harche** [comment]
The last run for the job `pull-kubernetes-node-crio-e2e` with the latest changes seem to have passed for crio + systemd configuration https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98374/pull-kubernetes-node-crio-e2e/1394975602431234048/

Putting it here explicitly because that job doesn't show up in the ci job list.

**Comment by dims** [comment]
Great news @harche !

**Comment by kolyshkin** [comment]
I have included the commit from this PR to https://github.com/kubernetes/kubernetes/pull/102147, as it fixes the regression caused by rc94/rc95 update.

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 3]
```diff
@@ -594,15 +535,28 @@ func (m *cgroupManagerImpl) Update(cgroupConfig *CgroupConfig) error {
 	}
 
 	if unified {
-		if err := setResourcesV2(libcontainerCgroupConfig); err != nil {
-			return fmt.Errorf("failed to set resources for cgroup %v: %v", cgroupConfig.Name, err)
+		if err := propagateControllers(libcontainerCgroupConfig.Path); err != nil {
+			return err
```
Comment (kolyshkin) [inline pkg/kubelet/cm/cgroup_manager_linux.go:539]: @odinuge why is this needed? runc's cgroup manager does the same thing in `Apply()` once, and doing it on every update/set seems excessive.

Comment (odinuge) [reply pkg/kubelet/cm/cgroup_manager_linux.go:539]: > why is this needed?

I just want to keep it as close to the original code as possible to avoid regressions or new issues. I think we instead can improve it later when we have better test coverage, and can do more testing.

> runc's cgroup manager does the same thing in Apply() once, and doing it on every update/set seems excessive.

Yeah, I know. I can remove it tho., but I would instead like to do it in a follow up.

There is a lot of duplicate work in both k8s and libcontainer, especially when using systemd. I do think we can boost performance quite a bit by improving it (both in cpu cycles in kubelet and in locks taken by the kernel cgroup code), but that is outside the scope of this imo.

Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:539]: Right; I missed it was in `setResourcesV2`. Can and should be addressed later.

**Comment by kolyshkin** [comment]
I think this one can be closed in favor of  #102147 (which includes a version of commit from this PR for runc rc94+).

**Comment by odinuge** [comment]
/close

In favor of #102147
